leap_c/examples/hvac/controller.py

```diff
@@ -0,0 +1,542 @@
+    
+    # Adjust layout and add summary stats
+    plt.tight_layout()
+    _add_summary_stats(fig, u, ctx)
+    
+    # Save figure if path provided
+    if save_path:
+        plt.savefig(save_path, dpi=300, bbox_inches="tight")
+        print(f"Figure saved to: {save_path}")
+    
+    return fig
+
+
+def plot_simulation(time: np.ndarray, obs: np.ndarray, action: np.ndarray) -> plt.Figure:
+    """
+    Plot simulation results in a figure with four vertically stacked subplots.
+
+    Args:
+        time: Time array
+        obs: Observation data
+        action: Control action array
+
+    Returns:
+        matplotlib Figure object
+    """
+    quarter_hours, day, Ti, Th, Te, Ta_forecast, solar_forecast, price_forecast = decompose_observation(obs=obs)
+    
+    # Convert temperatures to Celsius for plotting
+    Ti_celsius = convert_temperature(Ti, "kelvin", "celsius")
+    Th_celsius = convert_temperature(Th, "kelvin", "celsius")
+    Te_celsius = convert_temperature(Te, "kelvin", "celsius")
+    Ta_celsius = convert_temperature(Ta_forecast[:, 0], "kelvin", "celsius")
+    
+    Ti_lower, Ti_upper = set_temperature_limits(quarter_hours)
+    Ti_lower_celsius = convert_temperature(Ti_lower, "kelvin", "celsius")
+    Ti_upper_celsius = convert_temperature(Ti_upper, "kelvin", "celsius")
+    
+    qh = action
+    solar = solar_forecast[:, 0]
+    price = price_forecast[:, 0]
+    
+    # Create base plot
+    fig, axes = _create_base_plot()
+    
+    # Plot each subplot using helper functions
+    _plot_temperature_subplot(axes[0], time, Ti_celsius, Te_celsius, Ti_lower_celsius, Ti_upper_celsius)
+    _plot_heater_subplot(axes[1], time, Th_celsius)
+    _plot_disturbance_subplot(axes[2], time, Ta_celsius, solar)
+    _plot_control_subplot(axes[3], time, qh, price)
+    
+    # Adjust layout and add summary stats
+    plt.tight_layout()
+    _add_summary_stats(fig, qh)
+    
+    return fig
+
+
+if __name__ == "__main__":
+    horizon_hours = 24
+    N_horizon = horizon_hours * 4  # 4 time steps per hour
+
+    start_time = pd.Timestamp('2021-01-01 00:00:00+0100', tz='UTC+01:00')
+    env = StochasticThreeStateRcEnv(
+        step_size=900.0,  # 15 minutes in seconds
+        horizon_hours=horizon_hours,
+        start_time=start_time,
+    )
+
+    controller = HvacController(
+        N_horizon=N_horizon,
+        diff_mpc_kwargs={
+            "export_directory": Path("hvac_mpc_export"),
+        },
+    )
+
+    n_steps = 1 * 24 * 4  # days * hours * 4 time steps per hour
+
+    obs, info = env.reset()
+
+    obs = np.tile(obs, (n_steps + 1, 1))
+    action = np.zeros((n_steps, 1), dtype=np.float32)
+    time = []
+
+    for k in range(n_steps):
+        time.append(info["time_forecast"][0])
+        _, action[k] = controller.forward(obs=obs[k, :].reshape(1, -1))
+        obs[k+1, :], _, _, _, info, _ = env.step(action=action[k])
+
+    time = np.array(time)
+    obs = obs[:-1, :]
+
+    plot_simulation(time, obs, action)
```

Comment:
The plot looks great! But I think it would look even better with x limits corresponding to the simulation time.